numba/np/arrayobj.py

```diff
@@ -1907,6 +1907,73 @@ def impl(start, stop, num=50):
     return impl
 
 
+@overload(np.geomspace)
+def numpy_geomspace(start, stop, num=50):
+    if not isinstance(start, types.Number):
+        msg = 'The argument "start" must be a number'
+        raise errors.TypingError(msg)
+
+    if not isinstance(stop, types.Number):
+        msg = 'The argument "stop" must be a number'
+        raise errors.TypingError(msg)
+
+    if not isinstance(num, (int, types.Integer)):
+        msg = 'The argument "num" must be an integer'
+        raise errors.TypingError(msg)
+
+    if any(isinstance(arg, types.Complex) for arg in [start, stop]):
+        def impl(start, stop, num=50):
+            if start == 0 or stop == 0:
+                raise ValueError('Geometric sequence cannot include zero')
+            both_negative = (np.sign(start) == -1) & (np.sign(stop) == -1)
+            out_sign = 1
+            if (start.real == 0) & (stop.real == 0):
+                start = start.imag
+                stop = stop.imag
+                out_sign = 1j
+            if both_negative:
+                start = -start
+                stop = -stop
+                out_sign = -out_sign
+            start = np.complex64(start)
+            stop = np.complex64(stop)
```

Comment:
This is truncating start and stop to `complex64`. In my system, the default type for complex values is `np.complex128` and the following inputs gives different results:

```
from numba import njit
import numpy as np

@njit
def foo(start, stop, num):
    return np.geomspace(start, stop, num=num)

start = 1j
stop = np.complex64(16j)
num = 3
print(foo(start, stop, num))
print(np.geomspace(start, stop, num=num))

# [0. +1.j         0. +4.00000026j 0.+16.j        ]
# [0. +1.j 0. +4.j 0.+16.j]
```

At compile time you can determine the resulting dtype (use `np.promote_types` for that) and use cast start/stop using this resulting dtype. Code should look something like this:

```python
    if any(isinstance(arg, types.Complex) for arg in [start, stop]):
        from numba.np.numpy_support import as_dtype, from_dtype
        result_dtype = from_dtype(np.promote_types(as_dtype(start), as_dtype(stop)))
        def impl(start, stop, num=50):
            ...
            start = result_dtype(start)
            stop = result_dtype(stop)
            ...
```